jwst/undersampling_correction/undersampling_correction.py

```diff
@@ -61,44 +66,79 @@ def flag_pixels(data, gdq, signal_threshold):
 
     Returns
     -------
-    gdq : int, 4D array
+    new_gdq : int, 4D array
         updated group dq array
     """
     n_ints, n_grps, n_rows, n_cols = gdq.shape
-    num_pix = n_cols * n_rows
-
-    lowest_exc_1d = np.zeros(num_pix) + n_grps
-
-    for ii_int in range(n_ints):
-        for ii_grp in range(n_grps):
-            data_1d = data[ii_int, ii_grp, :, :].reshape(num_pix)  # vectorize slice
-            gdq_1d = gdq[ii_int, ii_grp, :, :].reshape(num_pix)
-
-            wh_not_dnu = np.logical_not(gdq_1d & dqflags.group['DO_NOT_USE'])
-
-            # In the current group for all ramps, locate pixels that :
-            #  a) exceed the signal_threshold, and
-            #  b) have not been previously flagged as an exceedance, and
-            #  c) were not flagged in an earlier step as DO_NOT_USE
-            wh_exc_1d = np.where((data_1d > signal_threshold) &
-                                 (lowest_exc_1d == n_grps) & wh_not_dnu)
-
-            # ... and mark those pixels, as current group is their first exceedance
-            if len(wh_exc_1d[0] > 0):  # For ramps previously unflagged ...
-                lowest_exc_1d[wh_exc_1d] = ii_grp
-
-    # Flag current and subsequent groups
-    lowest_exc_2d = lowest_exc_1d.reshape((n_rows, n_cols))
-    for ii_int in range(n_ints):
-        for ii_grp in range(n_grps):
-            wh_set_flag = np.where(lowest_exc_2d == ii_grp)
-
-            # set arrays of components
-            yy = wh_set_flag[0]
-            xx = wh_set_flag[1]
-
-            gdq[ii_int, ii_grp:, yy, xx] = \
-                np.bitwise_or(gdq[ii_int, ii_grp:, yy, xx], dqflags.group['UNDERSAMP']
-                              | dqflags.group['DO_NOT_USE'])
-
-    return gdq
+    ncols = data.shape[3]
+    nrows = data.shape[2]
+
+    new_gdq = gdq.copy()   # Updated gdq
+
+    # Flag all exceedances with UNDERSAMP and NO_NOT_USE
+    undersamp_pix = (data > signal_threshold) == (gdq != DNU)
+    new_gdq[undersamp_pix] = np.bitwise_or(new_gdq[undersamp_pix], UNSA)
+    new_gdq[undersamp_pix] = np.bitwise_or(new_gdq[undersamp_pix], DNU)
+
+    # Reset groups previously flagged as DNU
```

Comment:
Not sure I understand why this needs to be done. `new_gdq` and `gdq_orig` are both copies of `gdq`, so they should be identical. Then `new_gdq` has additional pixels flagged as DNU+UNSA, but only if they were not already flagged as DNU. So any pixels originally flagged (on input) as DNU should still be there in `new_gdq`, won't they? So why do they need to be copied over again from `gdq_orig`, which is equal to `gdq`?  Maybe I'm missing something.

Reply:
Ah right, I understand about the neighbors and you're right that we don't want to flag a neighbor if it was already set to DNU. But isn't this operation taking place *before* the attempt to flag neighbors? I would think it would need to be done *after* the neighbors get flagged, i.e. to undo any damage caused by the neighbor flagging.

Reply:
As the code is now, it us done both before any neighbors are flagged (which I agree can be deleted) and again after neighbors are flagged (which should remain).